### DISCUSSION TITLE
Missing replicasets related metrics

### DISCUSSION THREAD

**Discussion by kjurasze** [percona/mongodb_exporter#933] [Q&A] [answered]
Hello,
could you please give an advice why replicasets metrics (eg. `mongodb_mongod_replset_my_state` ) might be missing ? Is the exporter ready for working with sharded clusters ?

Currently mongodb_mongod_replset_my_state returns only one, without set and with a value 6 which is UNKNOWN. The exporter itself isn't throwing any errors. 
Running exporter with a debug logging level is giving some output, eg.
```
"replSetGetStatus": {
  "failed": 701,
  "total": 701
},
```

which is strange, since the user has sufficient priviliges (tested manually via mongosh). Also having look at the certain rs logs there're successful authentications of that user without any errors.
Running the exporter with a different sets of args, eg, collector.replicasetstatus only (combined with the other collectors), no-direct-connect, global-conn-pool etc. didn’t have any effect.

Tested with 0.39, 0.40 and 0.41.0 versions of the exporter.
MongoDB: 7.0.14-8 (metrics were missing also on 6.x)
Actual args:
```
       --web.listen-address=:9216
       --discovering-mode
       --collector.shards
       --collector.replicasetstatus
       --collector.diagnosticdata
       --collector.dbstats
       --collector.dbstatsfreestorage
       --collector.collstats
       --collector.indexstats
       --collector.topmetrics
       --collector.profile
       --collector.currentopmetrics
       --compatible-mode
       --mongodb.global-conn-pool
       --no-mongodb.direct-connect
```

`MONGODB_URI` is a non-TLS URI to a mongos service, `mongodb:// <user> : <pass> @ <mongos-service-address> `

Best regards, Krzysztof

**Comment by BupycHuk** [comment]
mongos doesn't return replicaset metrics

**Comment by BupycHuk** [comment, marked as answer]
Hi, mongos doesn't return replicaset related metrics. Please run exporter against mongod instance

**Comment by kjurasze** [comment]
Thank you for a quick response. I'm still struggling to get the data from a cluster, I've tried with those URIs
- `mongodb:// <user> : <pass> @ <mongos-service-address>, <rs1>, <rs2> ...` 
- `mongodb:// <user> : <pass> @ <rs1>,mongodb:// <user> : <pass> @ <rs2> ...` 
- `mongodb:// <user> : <pass> @ <mongos-service-address>, <rs1-address>/?replicaSet=rs1, <rs2-address>/?replicaSet=rs2 ...` 
and best result I got were metrics for only one (random) replicaset. Could you provide an examplary URI for a cluster consisting of a multiple replicasets ?

**Comment by kjurasze** [comment]
@BupycHuk do you have idea why exporter would export only one replicaset data at the time ?
Example:
![image](https://github.com/user-attachments/assets/1d1f301a-71ec-4526-a03e-672296a27472)
used metrics in this chart: `mongodb_mongod_replset_oplog_head_timestamp`, `mongodb_mongod_replset_oplog_tail_timestamp`, `mongodb_mongod_replset_my_state`.

**Reply by BupycHuk** [reply]
initially we designed mongodb_exporter to run it per each node, that's why we didn't expected users to use one mongodb_exporter to monitor whole cluster. Hovewer there are a few improvements by @adnull to monitor multiple instances. we haven't tested it properly yet.
